examples/websocket_echo_server/main.pony

```diff
@@ -0,0 +1,124 @@
+use "crypto"
+use "encode/base64"
+use "net"
+
+use "http_server"
+use ws = "websocket"
+
+actor Main
+  new create(env: Env) =>
+    let host = "localhost"
+    let port = "9999"
+    let limit: USize = 9000
+    let server = Server(
+      TCPListenAuth(env.root),
+      SimpleServerNotify,  // notify for server lifecycle events
+      BackendMaker   // factory for session-based application backend
+      where config = ServerConfig( // configuration of Server
+        where host' = host,
+              port' = port,
+              max_concurrent_connections' = limit)
+    )
+
+class SimpleServerNotify
+  fun ref listening(server: Server ref) =>
+    None
+
+  fun ref not_listening(server: Server ref) =>
+    None
+
+  fun ref closed(server: Server ref) =>
+    None
+
+class BackendMaker is HandlerFactory
+  fun apply(session: Session): Handler^ =>
+    RequestHandler.create(session)
+
+class RequestHandler is Handler
+  let _session: Session
+  var _response_builder: ResponseBuilder
+
+  new ref create(session: Session) =>
+    _session = session
+    _response_builder = Responses.builder()
+
+  fun ref apply(request: Request val, request_id: RequestID) =>
+    if (request.method() isnt GET) or (request.version() < HTTP11) then
+      let body = "Invalid Method or HTTP version"
+      this._send_err(request_id, StatusBadRequest, body)
+    else 
+      try
+        let upgrade_header = request.header("Upgrade") as String
+        if upgrade_header != "websocket" then
+          error
+        end
+        let conn_header = request.header("Connection") as String
+        if conn_header != "Upgrade" then
+          error
+        end
+        // calculate Sec-Websocket-Accept from Sec-Websocket-Key
+        let ws_key = request.header("Sec-WebSocket-Key") as String
+        let sha1_digest = Digest.sha1()
+        sha1_digest.append(ws_key)?
+        sha1_digest.append("258EAFA5-E914-47DA-95CA-C5AB0DC85B11")?
+        let hash: Array[U8] val = sha1_digest.final()
+        let sec_ws_accept: String val = Base64.encode(hash)
+
+        let ws_version = request.header("Sec-WebSocket-Version") as String
+        if ws_version.u64()? != 13 then
+          error
+        end
+        _session.send_raw(
+          _response_builder.set_status(StatusSwitchingProtocols)
+            .add_header("Upgrade", "websocket")
+            .add_header("Connection", "Upgrade")
+            .add_header("Sec-WebSocket-Accept", sec_ws_accept)
+            .finish_headers()
+            .build(),
+          request_id
+        )
```

Comment:
Should this be integrated be integrated with the `_session.upgrade` method so that this "Switching Protocols" response will definitely happen (without the user needing to remember to implement this)?

I think it would be better to make this all in one call rather than asking the protocol switching handler to construct a raw response.

Reply:
Are websockets the only reason one would upgrade the session, or might there be other reasons and the header would look different?

Reply:
Upgrade is a general part of the HTTP protocol, so it isn't websocket specific.

Reply:
I thought about integrating sending a 101 Response with in the `Session.upgrade_protocol` function. I initially voted against it, because i wanted to keep the library as raw and low-level as possible. While nearly all protocol upgrades will be used for Websockets, I'd rather not block the path for upgrading to other protocols. And people might want to add special headers to their response and we have multiple API styles used in this library for sending responses, so a generic upgrade_protocol method that does all the upgrading boilerplate while still being generic enough so people can change the response as they see fit, didn't seem worthwile to me.

I'd suggest to move the discussion about a more higher-level API for this to new Issue if need be.